twenty-first/src/shared_math/polynomial.rs

```diff
@@ -1074,10 +1067,67 @@ impl<PFElem: FiniteField> Sub for Polynomial<PFElem> {
     }
 }
 
+impl<BFieldElement: FiniteField> Polynomial<BFieldElement> {
+    // Division in 𝔽_p[X], not 𝔽_{p^e} ≅ 𝔽[X]/p(x).
+    pub fn xgcd(
+        mut x: Polynomial<BFieldElement>,
+        mut y: Polynomial<BFieldElement>,
+    ) -> (
+        Polynomial<BFieldElement>,
+        Polynomial<BFieldElement>,
+        Polynomial<BFieldElement>,
+    ) {
+        let mut a_factor = Polynomial::new(vec![BFieldElement::one()]);
+        let mut a1 = Polynomial::new(vec![BFieldElement::zero()]);
+        let mut b_factor = Polynomial::new(vec![BFieldElement::zero()]);
+        let mut b1 = Polynomial::new(vec![BFieldElement::one()]);
+
+        while !y.is_zero() {
+            let (quotient, remainder): (Polynomial<BFieldElement>, Polynomial<BFieldElement>) =
+                x.clone().divide(y.clone());
+            let (c, d) = (
+                a_factor - quotient.clone() * a1.clone(),
+                b_factor.clone() - quotient * b1.clone(),
+            );
+
+            x = y;
+            y = remainder;
+            a_factor = a1;
+            a1 = c;
+            b_factor = b1;
+            b1 = d;
+        }
+
+        // The result is valid up to a coefficient, so we normalize the result,
+        // to ensure that x has a leading coefficient of 1.
+        // TODO: What happens if x is zero here, can it be?
+        let lc = x.leading_coefficient().unwrap();
+        let scale = lc.inverse();
+        (
+            x.scalar_mul(scale),
+            a_factor.scalar_mul(scale),
+            b_factor.scalar_mul(scale),
+        )
+    }
+}
+
 impl<PFElem: FiniteField> Polynomial<PFElem> {
     pub fn degree(&self) -> isize {
         degree_raw(&self.coefficients)
     }
+
+    pub fn formal_derivative(&self) -> Self {
+        let coefficients = self
+            .clone()
+            .coefficients
+            .iter()
+            .enumerate()
+            .map(|(i, &coefficient)| PFElem::new_from_usize(&coefficient, i) * coefficient)
```

Comment:
I suggest changing the trait for `new_from_usize` to not take `&self` an argument.

Reply:
I agree. But we can do that in a separate commit/PR.